src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs

```diff
@@ -36,21 +33,68 @@ public static class TopologicalSortUtility
             return sorted;
         }
 
-        private static int GetParentCount(List<PackageDependencyInfo> packages, string id)
+        private static void Sort(PackageInfo[] packages, int start)
+        {
+            // Update parent counts
+            for (var i = start; i < packages.Length; i++)
+            {
+                var package = packages[i].Package;
+                // Mutating stuct in-place, need to index rather than use enumerator
+                packages[i].ParentCount = GetParentCount(packages, package.Id, start);
+            }
+
+            Array.Sort(packages, start, packages.Length - start, _comparer);
+        }
+
+        private static int GetParentCount(PackageInfo[] packages, string id, int start)
         {
-            int count = 0;
+            var parentCount = 0;
 
-            foreach (var package in packages)
+            for (var i = start; i < packages.Length; i++)
             {
-                if (package.Dependencies != null
-                    && package.Dependencies.Any(dependency =>
-                        string.Equals(id, dependency.Id, StringComparison.OrdinalIgnoreCase)))
+                var deps = packages[i].Package.Dependencies;
+                var dependencies = deps as PackageDependency[] ?? deps.ToArray();
```

Comment:
> For instance, here at the line 20 ToArray statement may be more expensive than list vs. array for-loop perf gain.

Previously it was doing `.ToList()` and `.ToArray()` is approximately equivalent. 

There is only `1` array allocation for `N` items; whereas previously it would be `N + (1 + N) / 2` allocations as it allocates per loop. Will put together a benchmark.